metagraph/src/annotation/binary_matrix/row_diff/row_diff.cpp

```diff
@@ -1,118 +1,154 @@
-    f.read(reinterpret_cast<char *>(&num_columns_), 8);
-    f.read(reinterpret_cast<char *>(&num_relations_), 8);
-    diffs_.load(f);
-    boundary_.load(f);
-    terminal_.load(f);
-    sdsl::util::init_support(sboundary_, &boundary_);
-    return true;
-}
-
-void RowDiff::serialize(std::ostream &f) const  {
-    f.write(reinterpret_cast<const char *>(&num_columns_), 8);
-    f.write(reinterpret_cast<const char *>(&num_relations_), 8);
-    diffs_.serialize(f);
-    boundary_.serialize(f);
-    terminal_.serialize(f);
-}
-
-void RowDiff::serialize(const std::string &name) const {
-    std::ofstream f(name, ios::binary);
+template <class BaseMatrix>
+void RowDiff<BaseMatrix>::serialize(const std::string &filename) const {
+    std::ofstream f(filename, ios::binary);
     serialize(f);
     f.close();
 }
 
-bool RowDiff::load(const std::string &name) {
-    std::ifstream f(name, ios::binary);
+template <class BaseMatrix>
+bool RowDiff<BaseMatrix>::load(const std::string &filename) {
+    std::ifstream f(filename, ios::binary);
     bool result = load(f);
     f.close();
-    return result;
-}
 
-Vector<uint64_t> RowDiff::get_diff(uint64_t node_id) const {
-    assert(boundary_[boundary_.size() - 1] == 1);
-    Vector<uint64_t> result;
-    uint64_t start_idx = node_id == 0 ? 0 : sboundary_.select(node_id) + 1;
+    load_terminal(anchors_filename_, &terminal_);
 
-    while (boundary_[start_idx] == 0) {
-        result.push_back(diffs_[start_idx - node_id]);
-        start_idx++;
-    }
     return result;
 }
 
-void RowDiff::merge(Vector<uint64_t> *result, const Vector<uint64_t> &diff2) {
-    if (diff2.empty()) {
-        return;
-    }
-    Vector<uint64_t> diff1;
-    std::swap(*result, diff1);
-    uint64_t idx1 = 0;
-    uint64_t idx2 = 0;
-    while (idx1 < diff1.size() && idx2 < diff2.size()) {
-        if (diff1[idx1] == diff2[idx2]) {
-            idx1++;
-            idx2++;
-        } else if (diff1[idx1] < diff2[idx2]) {
-            result->push_back(diff1[idx1]);
-            idx1++;
-        } else {
-            result->push_back(diff2[idx2]);
-            idx2++;
+template
+class RowDiff<ColumnMajor>;
+
+bool build_successor(const graph::DBGSuccinct &graph,
+                     const std::string &outfbase,
+                     uint32_t max_length,
+                     uint32_t num_threads) {
+    bool must_build = false;
+    for (const auto &suffix : { ".succ", ".pred", ".pred_boundary", ".terminal" }) {
+        if (!std::filesystem::exists(outfbase + suffix)) {
+            common::logger->trace(
+                    "Building and writing successor/predecessor/terminal files to {}.*",
+                    outfbase);
+            must_build = true;
+            break;
         }
     }
-    while (idx1 < diff1.size()) {
-        result->push_back(diff1[idx1]);
-        idx1++;
+    if (!must_build) {
+        common::logger->trace("Using existing pred/succ/terminal files in {}.*", outfbase);
```

Comment:
```suggestion
        common::logger->trace("Using existing pred / succ / terminal files in {}.*", outfbase);
```
otherwise looks like a path